python/triton/compiler/backends/cuda.py

```diff
@@ -184,6 +184,8 @@ def make_ptx(src, metadata, opt, capability):
             ptx_version = ptx_get_version(cuda_version)
         ptx_version = f'{ptx_version//10}.{ptx_version%10}'
         ret = re.sub(r'\.version \d+\.\d+', f'.version {ptx_version}', ret, flags=re.MULTILINE)
+        # Remove the debug flag that prevents ptxas from optimizing the code
+        ret = re.sub(r",\s*debug|debug,\s*", "", ret)
```

Comment:
isn't this the expected behavior of debug? Where does the ptx version comes from in this case and why does it has debug in it?

Reply:
There are historical reasons behind this problem. Triton carries debug information through all its passes, and when generating the PTX code, LLVM adds a 'debug' flag. However, our goal is not to generate a binary with "full" debug information, but rather one with just line mapping. NVIDIA's compilers and assemblers, particularly the external versions, cannot emit full debugging information when optimizations are enabled. They don't even emit standard inlining information. Therefore, to produce optimized code that includes some lightweight debug information, they provide a solution that involves inserting partial line information into the generated binary, as indicated by the flags `-lineinfo -O3`.

Reply:
Three scenarios:
- If you want full device debugging information, `-O3 -G` or `-O3 -device-debug` is not allowed in both `nvcc` and `ptxas`.
- If you want nvidia's custom inlining information, `--optimization-info -lineinfo -O3` can be passed through `nvcc` but it is not supported in `ptxas` AFAIK.
- If you only want minimum line mapping information, `-lineinfo -O3` can be used in both `nvcc` and `ptxas`.